Services/Cache.cs

```diff
@@ -55,89 +62,154 @@ public class Cache : ICache
             catch (ResourceNotFoundException)
             {
-                this.log.Info($"{CACHE_COLLECTION_ID}:{CACHE_KEY} not found.", () => {});
+                this.log.Info($"Cache get: cache {CACHE_COLLECTION_ID}:{CACHE_KEY} was not found", () => { });
                 return new CacheValue { Tags = new HashSet<string>(), Reported = new HashSet<string>() };
             }
         }
 
-        public async Task RebuildCacheAsync(bool force = false)
+        public async Task<bool> RebuildCacheAsync(bool force = false)
         {
             int retry = 5;
             while (true)
             {
-                HashSet<string> reportedNames = null;
-                DeviceTwinName twinNames = null;
                 ValueApiModel cache = null;
-                string etag = null;
 
+                // Retrieve cache to check if rebuilding is necessary
                 try
                 {
                     cache = await this.storageClient.GetAsync(CACHE_COLLECTION_ID, CACHE_KEY);
                 }
                 catch (ResourceNotFoundException)
                 {
-                    this.log.Info($"{CACHE_COLLECTION_ID}:{CACHE_KEY} not found.", () => {});
+                    this.log.Info($"Cache rebuilding: cache {CACHE_COLLECTION_ID}:{CACHE_KEY} was not found", () => { });
                 }
 
-                bool needBuild = this.NeedBuild(force, cache);
+                var needBuild = this.NeedBuild(force, cache);
                 if (!needBuild)
                 {
-                    return;
+                    this.log.Warn("Cache rebuilding skipped", () => { });
+                    return false;
                 }
 
+                // Build the cache content
+                DeviceTwinName twinNames;
                 try
                 {
-                    var twinNamesTask = this.iotHubClient.GetDeviceTwinNamesAsync();
-                    var reportedNamesTask = this.simulationClient.GetDevicePropertyNamesAsync();
-                    Task.WaitAll(twinNamesTask, reportedNamesTask);
+                    var twinNamesTask = this.GetValidNamesAsync();
+                    var simulationNamesTask = this.simulationClient.GetDevicePropertyNamesAsync();
+                    Task.WaitAll(twinNamesTask, simulationNamesTask);
                     twinNames = twinNamesTask.Result;
-                    reportedNames = reportedNamesTask.Result;
-                    reportedNames.UnionWith(twinNames.ReportedProperties);
+                    twinNames.ReportedProperties.UnionWith(simulationNamesTask.Result);
                 }
                 catch (Exception)
                 {
-                    this.log.Info($"retry{retry}: IothubManagerService and SimulationService  are not both ready,wait 10 seconds ", () => {});
                     if (retry-- < 1)
                     {
-                        return;
+                        this.log.Warn("Cache rebuilding failed due to out of retry times", () => { });
```

Comment:
log.error instead of warning - this failing breaks the application.

Reply:
Log removed since it will continue retrying forever.